File: Programmers/Python3/Level 3/숫자 게임/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/12987

from bisect import bisect_right


def solution(A, B):  # 첫번째 코드
    A.sort()
    B.sort()
    used = [False] * len(B)
    win = 0
    # A의 카드보다 큰 수 중에서 가장 차이가 적은 카드를 B에서 이진탐색을 통해 탐색
    for card in A:
        index = bisect_right(B, card)
        while index < len(B):
            if not used[index]:
                win += 1
                used[index] = True
                break
            index += 1
    return win


def solution(A, B):  # 두번째 코드
    A.sort()
    B.sort()
    win = 0
    a_index = 0
    # A와 B의 작은값부터 비교를 해가는 방식
    for card in B:
        if card > A[a_index]:
            a_index += 1
            win += 1
    return win

File: Programmers/Javascript/Level 3/숫자 게임/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/12987?language=javascript

function solution(A, B) {
  A.sort((a, b) => a - b);
  B.sort((a, b) => a - b);

  let answer = 0,
    remainCrew = 0;

  for (let idx = 0; idx < A.length; idx++) {
    if (remainCrew === B.length) {
      break;
    }
    // B에 남은 크루중 A의 크루를 이길수 있는 가장 작은 B크루를 찾음
    for (let crew = remainCrew; crew < B.length; crew++) {
      remainCrew += 1;
      if (B[crew] > A[answer]) {
        answer += 1;
        break;
      }
    }
  }
  return answer;
}
